src/client/opamArg.ml

```diff
@@ -782,62 +774,414 @@ let opamlist_columns =
+    let doc = update_doc_w_cli doc ~cli validity in
+    let doc = Arg.info ?docs:section ~docv:value ~doc flags in
+    let check elem =
+      check_cli_validity cli validity
+        ~cond:(fun c -> c && default != elem) elem flags
+    in
+    term_cli_check ~check Arg.(opt_all ?vopt kind default & doc)
+
+  let mk_vflag ~cli ?section default flags =
+    let flags = List.map (fun (v,c,f,d) -> contented_validity v c, f, d) flags in
+    let info_flags =
+      List.map (fun (validity, flag, doc) ->
+          let doc = update_doc_w_cli doc ~cli validity in
+          validity.content, Arg.info ?docs:section flag ~doc)
+        flags
+    in
+    let check elem =
+      match
+        OpamStd.List.find_opt (fun (validity, _, _) ->
+            validity.content = elem)
+          flags
+      with
+      | Some (validity, flags, _) ->
+        check_cli_validity cli validity (elem_of_vr elem) flags
+      | None -> `Ok (elem_of_vr elem)
+    in
+    term_cli_check ~check Arg.(vflag (`Valid default) info_flags)
+
+  let mk_vflag_all ~cli ?section ?(default=[]) flags =
+    let flags = List.map (fun (v,c,f,d) -> contented_validity v c, f, d) flags in
+    let info_flags =
+      List.map (fun (validity, flag, doc) ->
+          let doc = update_doc_w_cli doc ~cli validity in
+          validity.content, Arg.info ?docs:section flag ~doc)
+        flags
+    in
+    let check selected =
+      let neww, rm =
+        List.fold_left (fun (neww,rm) elem ->
+            match OpamStd.List.find_opt (fun (validity, _, _) ->
+                validity.content = elem) flags with
+            | Some ({ removed = None ; valid = c; _ }, flags, _)
+              when cond_new cli c ->
+              (flags, c)::neww, rm
+            | Some ({ removed = Some (removal, instead); _ }, flags, _)
+              when cond_removed cli removal ->
+              neww, (flags, (removal, instead))::rm
+            | _ -> neww,rm) ([],[]) selected
+      in
+      match neww, rm with
+      | [], [] -> `Ok (List.map elem_of_vr selected)
+      | [flags, c], [] ->
+        error_new_flag cli c flags
+      | [], [flags, (c, instead)] ->
+        error_rm_flag cli c instead flags
+      | _::_, []->
+        let options, clis = List.split neww in
+        let msg =
+          Printf.sprintf
+            "%s are only available since respectively versions %s of the opam \
+             CLI, but version %s has been requested."
+            (OpamStd.Format.pretty_list (List.map get_long_form options) )
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string clis))
+            (OpamCLIVersion.to_string cli)
+        in
+        `Error (false,msg)
+      | [], _::_->
+        let options, clis = List.split rm in
+        let clis = List.split clis |> fst in
+        let msg =
+          Printf.sprintf
+            "%s were removed in, respectively, %s versions of the opam CLI, \
+             but version %s has been requested."
```

Comment:
Likewise here,

```
            "%s were all removed by version %s of the opam CLI, \
             but version %s has been requested."
```
(this one _might_ be worth special-casing to display "were all _in_ version" if they all have the same version)